dvc/remote/gdrive/__init__.py

```diff
@@ -0,0 +1,243 @@
+from dvc.remote.base import RemoteBASE
+from dvc.config import Config
+from dvc.exceptions import DvcException
+from dvc.remote.gdrive.pydrive import (
+    RequestListFile,
+    RequestListFilePaginated,
+    RequestCreateFolder,
+    RequestUploadFile,
+    RequestDownloadFile,
+)
+from dvc.remote.gdrive.utils import FOLDER_MIME_TYPE
+
+logger = logging.getLogger(__name__)
+
+
+class GDriveURLInfo(CloudURLInfo):
+    @property
+    def netloc(self):
+        return self.parsed.netloc
+
+
+class RemoteGDrive(RemoteBASE):
+    scheme = Schemes.GDRIVE
+    path_cls = GDriveURLInfo
+    REGEX = r"^gdrive://.*$"
+    REQUIRES = {"pydrive": "pydrive"}
+    GDRIVE_USER_CREDENTIALS_DATA = "GDRIVE_USER_CREDENTIALS_DATA"
+    CREDENTIALS_FILE_PATH = "credentials.json"
+
+    def __init__(self, repo, config):
+        super(RemoteGDrive, self).__init__(repo, config)
+        self.no_traverse = False
+        self.path_info = self.path_cls(config[Config.SECTION_REMOTE_URL])
+        self.config = config
+        self.init_drive()
+
+    def init_drive(self):
+        self.gdrive_credentials_path = self.config.get(
+            Config.SECTION_REMOTE_KEY_FILE, None
+        )
+        if not self.gdrive_credentials_path:
+            raise DvcException(
+                "Google Drive settings file path is missed from config. "
+                "Learn more at "
+                "https://man.dvc.org/remote/add."
+            )
+        self.root_id = self.get_path_id(self.path_info, create=True)
+        self.cached_dirs, self.cached_ids = self.cache_root_dirs()
+
+    @on_exception(expo, DvcException, max_tries=8)
+    def execute_request(self, request):
+        try:
+            result = request.execute()
+        except Exception as exception:
+            retry_codes = ["403", "500", "502", "503", "504"]
+            if any(code in str(exception) for code in retry_codes):
+                raise DvcException("Google API request failed")
+            raise
+        return result
+
+    def list_drive_item(self, query):
+        list_request = RequestListFilePaginated(self.drive, query)
+        page_list = self.execute_request(list_request)
+        while page_list:
+            for item in page_list:
+                yield item
+            page_list = self.execute_request(list_request)
+
+    def cache_root_dirs(self):
+        cached_dirs = {}
+        cached_ids = {}
+        for dir1 in self.list_drive_item(
+            "'{}' in parents and trashed=false".format(self.root_id)
+        ):
+            cached_dirs.setdefault(dir1["title"], []).append(dir1["id"])
+            cached_ids[dir1["id"]] = dir1["title"]
+        return cached_dirs, cached_ids
+
+    @cached_property
+    def drive(self):
+        from pydrive.auth import GoogleAuth
+        from pydrive.drive import GoogleDrive
+
+        if os.getenv(RemoteGDrive.GDRIVE_USER_CREDENTIALS_DATA):
+            with open(self.CREDENTIALS_FILE_PATH, "w") as credentials_file:
```

Comment:
It's the multiple lines hack to allow test_data_cloud push/pull to real remote drive without manual user login into Google account to access GDrive.

The path where PyDrive searches for credentials file can be customized in settings.yaml file.
I'm going to think how to change this code to look right.

There is also Ivan's suggestion on this https://github.com/iterative/dvc/pull/2551#discussion_r341867620

Reply:
@MaxRis looks like it is the biggest concern and last big blocker (may be more tests would be great to have) - managing credentials. Since this is the external interface we'll be providing we need to get it right. Besides that it looks fragile, as we discussed, we should try to make it part of the regular DVC config flow.